src/swap/defi/rango.ts

```diff
@@ -0,0 +1,542 @@
+          exchangeInfo = asExchangeInfo(await exchangeInfoResponse.json())
+          exchangeInfoLastUpdate = now
+        } else {
+          // Error is ok. We just use defaults
+          const text: string = await exchangeInfoResponse.text()
+          log.warn(
+            `Error getting info server exchangeInfo. Using defaults... Error: ${text}`
+          )
+        }
+      } catch (e: any) {
+        log.warn(
+          'Error getting info server exchangeInfo. Using defaults...',
+          e.message
+        )
+      }
+    }
+
+    const fromSymbol =
+      rangoTokens?.[fromMainnetCode]?.[fromContractAddress.toLowerCase()]
+    const toSymbol =
+      rangoTokens?.[toMainnetCode]?.[toContractAddress.toLowerCase()]
+
+    if (fromSymbol === undefined || toSymbol === undefined) {
+      throw new Error('Rango could not proceed with the exchange')
+    }
+
+    if (exchangeInfo != null) {
+      const { rango } = exchangeInfo.swap.plugins
+      rangoServers = rango?.rangoServers ?? rangoServers
+    }
+
+    const swapParameters = {
+      apiKey: rangoApiKey,
+      from: assetToString({
+        blockchain: fromMainnetCode,
+        address: fromContractAddress,
+        symbol: fromSymbol
+      }),
+      to: assetToString({
+        blockchain: toMainnetCode,
+        address: toContractAddress,
+        symbol: toSymbol
+      }),
+      fromAddress: fromAddress,
+      toAddress: toAddress,
+      amount: nativeAmount,
+      disableEstimate: true,
+      slippage: DEFAULT_SLIPPAGE,
+      ...(referrerAddress !== undefined &&
+        referrerAddress !== '' &&
+        referrerFee !== undefined &&
+        referrerFee !== '' && { referrerAddress, referrerFee })
+    }
+
+    const params = makeQueryParams(swapParameters)
+
+    const swapResponse = await fetchWaterfall(
+      fetchCors,
+      rangoServers,
+      `basic/swap?${params}`,
+      {
+        headers
+      }
+    )
+
+    if (!swapResponse.ok) {
+      const responseText = await swapResponse.text()
+      throw new Error(`Rango could not fetch quote: ${responseText}`)
+    }
+
+    const swap = asSwapResponse(await swapResponse.json())
+    const { route, tx } = swap
+
+    if (swap.resultType !== 'OK') {
+      throw new Error(
+        `Rango could not proceed with the exchange. : ${swap.resultType}`
+      )
+    }
+
+    if (
+      route?.path == null ||
+      route.outputAmount === '' ||
+      tx == null ||
+      tx.txData == null
+    ) {
+      throw new Error('Rango could not proceed with the exchange')
+    }
+
+    const providers = route.path.map(p => p.swapper.title)
+
+    let preTx: EdgeTransaction | undefined
+    if (tx.type === 'EVM' && tx.approveData != null && tx.approveTo != null) {
+      const approvalData = tx.approveData
+
+      const spendInfo: EdgeSpendInfo = {
+        tokenId: null,
+        spendTargets: [
+          {
+            memo: approvalData,
+            nativeAmount: '0',
+            publicAddress: tx.approveTo
```

Comment:
For token approvals, the `spendTarget.publicAddress` should be the source token contract address.